java-checks/src/main/java/org/sonar/java/checks/helpers/IdentifierUtils.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2019 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.helpers;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Function;
+import javax.annotation.CheckForNull;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import static org.sonar.java.checks.helpers.ReassignmentFinder.getInitializerOrExpression;
+import static org.sonar.java.checks.helpers.ReassignmentFinder.getReassignments;
+
+public class IdentifierUtils {
+  private IdentifierUtils() {
+    // This class only contains static methods
+  }
+
+  public static <T> ValueResolution<T> getUniqueAssignmentValueResolution(ExpressionTree expression, Function<ExpressionTree,T> resolver) {
+    return valueResolution(expression, resolver, new ValueResolution<>());
+  }
+
+  private static <T> ValueResolution<T> valueResolution(ExpressionTree expression, Function<ExpressionTree,T> resolver, ValueResolution<T> valueResolution) {
```

Comment:
I think we could use `ConstantUtils.resolveAsConstant` as universal resolver and remove the need for `resolver` parameter. Users can cast the return value on the call site if needed.